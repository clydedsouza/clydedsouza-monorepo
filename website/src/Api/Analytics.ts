import mixpanel from 'mixpanel-browser'
import { PageTypes } from '../Components/Navigation/PageTypes'
import {
  ILinkClickedProps,
  MIXPANEL_DEV_PROJECT_ID,
  MIXPANEL_PROD_PROJECT_ID,
} from './IAnalytics'

export const initAnalyticsWithSuperProperties = () => {
  const mixpanelProjectId =
    process.env.NODE_ENV === 'production'
      ? MIXPANEL_PROD_PROJECT_ID
      : MIXPANEL_DEV_PROJECT_ID

  try {
    mixpanel.init(mixpanelProjectId)
    mixpanel.register({
      environment: process.env.NODE_ENV,
    })
  } catch {
    undefined
  }
}

export const sendPageViewEvent = (pageType: PageTypes) => {
  try {
    mixpanel.track('Page view', { page: pageType })
  } catch {
    undefined
  }
}

export const sendLinkClickedEvent = (props: ILinkClickedProps) => {
  try {
    mixpanel.track('Link clicked', {
      destination_url: props.link,
      link_type: props.type,
      page_location: props.location,
    })
  } catch {
    undefined
  }
}
