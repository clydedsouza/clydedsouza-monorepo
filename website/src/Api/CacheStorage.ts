import { PageTypes } from '../Components/Navigation/PageTypes'
import { IProjectData } from './IProjectData'

export const projectDataCacheMap: Map<PageTypes, IProjectData> = new Map()
