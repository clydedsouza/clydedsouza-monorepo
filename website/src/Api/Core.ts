import axios from 'axios'
import { PageTypes } from '../Types/PageTypes'
import { IProjectData } from '../Types/ProjectData'

const BASE_URL = 'https://api.clydedsouza.net/'

const getProjectApiUrl = (pageType: PageTypes) => {
  switch (pageType) {
    case PageTypes.Books:
      return `${BASE_URL}books.json`
    case PageTypes.Speaking:
      return `${BASE_URL}speaking.json`
    case PageTypes.Teaching:
      return `${BASE_URL}teaching.json`
    case PageTypes.Projects:
    default:
      return `${BASE_URL}pinned.json`
  }
}

export const getProjectData = (pageType: PageTypes): Promise<IProjectData> => {
  const request = getProjectApiUrl(pageType)
  return axios
    .get(request)
    .then((response) => {
      return response.data
    })
    .catch(() => {
      throw new Error('API error')
    })
}
