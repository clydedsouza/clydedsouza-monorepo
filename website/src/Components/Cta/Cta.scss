@import '../../Sass/variables';

.cta {
  background-color: rgba($contrast-accent, 0.1);
  border-radius: 6px;
  padding: 18px 12px;
  margin: 2vh 0;

  a {
    color: white;
    text-decoration: none;
    transition: all 0.2s linear 0.2s;
    border-bottom: 0;

    &:hover,
    &:focus {
      opacity: 0.8;
      border-bottom: 1px dashed white;
    }
  }

  @include for-tablets-down {
    display: none;
  }
}
