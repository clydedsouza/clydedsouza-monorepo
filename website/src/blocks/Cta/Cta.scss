@import "../../styles/variables";

.cta {
  background-color: rgba($contrast-accent, 0.1);
  border-radius: 6px;
  padding: 18px 12px;
  margin: 2vh 0;

  &[data-location="sidebar"] {
    display: block;

    @include for-tablets-down {
      display: none;
    }
  }

  &[data-location="footer"] {
    display: none;

    @include for-tablets-down {
      display: block;
    }
  }

  a {
    color: white;
    text-decoration: none;
    transition: all 0.2s linear 0.2s;
    border-bottom: 0;

    &:hover,
    &:focus {
      color: white;
      opacity: 0.8;
      border-bottom: 1px dashed white;
    }
  }
}
