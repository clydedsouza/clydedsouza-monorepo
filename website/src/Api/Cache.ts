import { PageTypes } from '../Types/PageTypes'
import { IProjectData } from '../Types/ProjectData'
import { getProjectData } from './Core'

const projectDataCacheMap: Map<PageTypes, Promise<IProjectData>> = new Map()

export const getCachedProjectData = (
  pageType: PageTypes
): Promise<IProjectData> => {
  if (!projectDataCacheMap.has(pageType)) {
    const getProjectDataPromise = getProjectData(pageType).catch((error) => {
      projectDataCacheMap.delete(pageType)
      return Promise.reject(error)
    })
    projectDataCacheMap.set(pageType, getProjectDataPromise)
  }

  return projectDataCacheMap.get(pageType)!
}
