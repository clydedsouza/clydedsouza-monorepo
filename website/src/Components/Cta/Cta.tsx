import { useEffect, useState } from 'react'
import { sendLinkClickedEvent } from '../../Api/Analytics'
import { getCachedProjectData } from '../../Api/Cache'
import { AnalyticsLinkType } from '../../Api/IAnalytics'
import { IProject } from '../../Api/IProjectData'
import Loader from '../Loader/Loader'
import { LoaderTypes } from '../Loader/LoaderTypes'
import { PageTypes } from '../Navigation/PageTypes'
import './Cta.scss'

interface ICtaProps {
  location?: string
}

function Cta({ location }: ICtaProps) {
  const [projectData, setProjectData] = useState<Partial<IProject>[]>([])
  const [isLoading, setIsLoading] = useState<boolean>(true)

  useEffect(() => {
    setIsLoading(true)
    getCachedProjectData(PageTypes.Cta)
      .then((apiResponse) => {
        setProjectData(apiResponse.data)
      })
      .catch(() => {
        setProjectData([])
      })
      .finally(() => {
        setIsLoading(false)
      })
  }, [])

  const CtaLink = (items: Partial<IProject>[]) => {
    if (items.length < 1) return
    const randomizedProject =
      items[Math.floor(Math.random() * (items.length - 1 - 0 + 1) + 0)]

    if (!randomizedProject.title || !randomizedProject.website) return
    return (
      <div className="cta" data-location={location}>
        <a
          href={randomizedProject.website}
          target="_blank"
          rel="noreferrer"
          onClick={() =>
            sendLinkClickedEvent({
              link: randomizedProject.website,
              type: AnalyticsLinkType.Cta,
              location: location,
            })
          }
        >
          {randomizedProject.title}
        </a>
      </div>
    )
  }

  return (
    <>
      {isLoading ? (
        <Loader variant={LoaderTypes.Inverse} />
      ) : projectData.length > 0 ? (
        CtaLink(projectData)
      ) : (
        <p>Couldn&lsquo;t load the data at this stage</p>
      )}
    </>
  )
}

export default Cta
