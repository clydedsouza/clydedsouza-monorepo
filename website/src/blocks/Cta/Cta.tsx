"use client";

import { sendLinkClickedEvent } from "@/api/Analytics";
import { getCachedProjectData } from "@/api/Cache";
import { AnalyticsLinkType } from "@/api/IAnalytics";
import { IProject } from "@/api/IProjectData";
import Loader from "@/blocks/Loader/Loader";
import { LoaderTypes } from "@/blocks/Loader/LoaderTypes";
import { useEffect, useState } from "react";
import { PageTypes } from "../Navigation/PageTypes";
import "./Cta.scss";

interface ICtaProps {
  location?: string;
}

function Cta({ location }: ICtaProps) {
  const [projectData, setProjectData] = useState<Partial<IProject>[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    setIsLoading(true);
    getCachedProjectData(PageTypes.Cta)
      .then((apiResponse) => {
        setProjectData(apiResponse.data);
      })
      .catch(() => {
        setProjectData([]);
      })
      .finally(() => {
        setIsLoading(false);
      });
  }, []);

  const CtaLink = (items: Partial<IProject>[]) => {
    if (items.length < 1) return;
    const randomizedProject =
      items[Math.floor(Math.random() * (items.length - 1 - 0 + 1) + 0)];

    if (!randomizedProject.title || !randomizedProject.website) return;
    return (
      <div className="cta" data-location={location}>
        <a
          href={randomizedProject.website}
          target="_blank"
          rel="noreferrer"
          onClick={() =>
            sendLinkClickedEvent({
              link: randomizedProject.website,
              type: AnalyticsLinkType.Cta,
              location: location,
            })
          }
        >
          {randomizedProject.title}
        </a>
      </div>
    );
  };

  return (
    <>
      {isLoading ? (
        <Loader variant={LoaderTypes.Inverse} />
      ) : projectData.length > 0 ? (
        CtaLink(projectData)
      ) : (
        <p>Couldn&lsquo;t load the data at this stage</p>
      )}
    </>
  );
}

export default Cta;
