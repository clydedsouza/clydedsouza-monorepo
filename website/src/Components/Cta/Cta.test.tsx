import {
  render,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import { emptyProject } from '../../../TestHelpers/testHelper'
import * as apiCacheModule from '../../Api/Cache'
import Cta from './Cta'

jest.mock('../../Api/Cache')

const ctaDataMockResponse = [
  {
    ...emptyProject,
    id: 'ai-tell-me-a-story',
    title: 'Purchase my book AI, Tell Me a Story from these platforms.',
    website: 'https://aitellmeastory.clydedsouza.net',
  },
  {
    ...emptyProject,
    id: 'mama-tell-me-a-story',
    title: 'Purchase my book Mama, Tell Me a Story from these platforms.',
    website: 'https://mamatellmeastory.clydedsouza.net',
  },
]

describe('Cta', () => {
  describe('given API return a single item', () => {
    beforeEach(() => {
      jest
        .spyOn(apiCacheModule, 'getCachedProjectData')
        .mockImplementationOnce(() =>
          Promise.resolve({
            app: {},
            data: [ctaDataMockResponse[0]],
          })
        )
    })

    it('should render cta', async () => {
      render(<Cta />)
      await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
      expect(document.body).toMatchSnapshot()
    })
  })

  describe.each([
    { title: '', website: '' },
    { title: 'test', website: '' },
    { title: '', website: 'test' },
  ])('given API return a single invalid item', ({ title, website }) => {
    beforeEach(() => {
      jest
        .spyOn(apiCacheModule, 'getCachedProjectData')
        .mockImplementationOnce(() =>
          Promise.resolve({
            app: {},
            data: [{ ...emptyProject, title, website }],
          })
        )
    })

    it(`should render cta when title is ${title} and website is ${website}`, async () => {
      render(<Cta />)
      await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
      expect(document.body).toMatchSnapshot()
    })
  })

  describe('given API returns multiple items', () => {
    beforeEach(() => {
      jest
        .spyOn(apiCacheModule, 'getCachedProjectData')
        .mockImplementationOnce(() =>
          Promise.resolve({
            app: {},
            data: [...ctaDataMockResponse],
          })
        )
    })

    it('should render cta', async () => {
      render(<Cta />)
      await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
      expect(screen.getByRole('link')).toHaveTextContent(/Purchase my book/)
      expect(screen.getByRole('link')).toHaveAttribute(
        'href',
        expect.stringContaining('tellmeastory.clydedsouza.net')
      )
    })
  })

  describe('given API returns no items', () => {
    beforeEach(() => {
      jest
        .spyOn(apiCacheModule, 'getCachedProjectData')
        .mockImplementationOnce(() =>
          Promise.resolve({
            app: {},
            data: [],
          })
        )
    })

    it('should render empty cta', async () => {
      render(<Cta />)
      await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
      expect(document.body).toMatchSnapshot()
    })
  })

  describe('given API returns an error', () => {
    beforeEach(() => {
      jest
        .spyOn(apiCacheModule, 'getCachedProjectData')
        .mockImplementationOnce(() => Promise.reject('Error'))
    })

    it('should render empty cta', async () => {
      render(<Cta />)
      await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
      expect(document.body).toMatchSnapshot()
    })
  })
})
