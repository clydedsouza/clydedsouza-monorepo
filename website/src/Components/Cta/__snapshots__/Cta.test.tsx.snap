// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cta given api return a single item should render cta 1`] = `
<body>
  <div>
    <div
      class="cta"
    >
      <a
        href="https://aitellmeastory.clydedsouza.net"
        rel="noreferrer"
        target="_blank"
      >
        Purchase my book AI, Tell Me a Story from these platforms.
      </a>
    </div>
  </div>
</body>
`;

exports[`Cta given api returns an error should render empty cta 1`] = `
<body>
  <div>
    <p>
      Couldn‘t load the data at this stage
    </p>
  </div>
</body>
`;

exports[`Cta given api returns no items should render empty cta 1`] = `
<body>
  <div>
    <p>
      Couldn‘t load the data at this stage
    </p>
  </div>
</body>
`;
